docs/unification.md

```diff
@@ -0,0 +1,344 @@
+* Otherwise if one or both terms are bound variables bound to another variable,
+  decompose with the binding of the bound variable and the other variable.
+* Otherwise, both terms are either free variables or variables bound to
+  constructors. If both terms are free variables, bind the lesser free variable
+  to the greater free variable according to the normal order of variables.
+* If one term is a free variable and the other is bound to a constructor, bind
+  the free variable to the constructor.
+* If both terms are variables bound to constructors, decompose with the
+  constructors.
+
+If one term is a variable and the other is a function-like pattern (but not a
+variable), consider the following cases:
+
+* If the variable is free, bind it to the pattern.
+* If the variable is bound to a variable, bind the variable it is bound to to
+  the pattern.
+* If the variable is bound to a constructor, decompose with the constructor and
+  the pattern.
+
+Constructors
+------------
+
+If the first term is an injective application and the second is the same
+symbol, decompose with each pair of children of each pattern.  If both terms
+are different constructors or overloaded symbols, fail.
+
+Injections
+----------
+
+If both terms are `inj`, and you are unifying `inj{s1, s}(p1)` and
+`inj{s2, s}(p2)`:
+
+* If s1 == s2, decompose with `p1` and `p2`.
+* If s1 < s2, decompose with `inj{s1, s2}(p1)` and `p2`.
+* If s2 < s1, symmetric of above
+* otherwise, fail
+
+If one term is `inj` and the other is a constructor, fail.
+
+Overloading
+-----------
+If one term is `inj{s1, s2}(p1)` where p1 is an overloaded application, and the
+other is `p2`, an overload of `p1`, or if both terms are injections applied
+to overloaded constructors, or if one term is an overloaded symbol and the
+other is an injection applied to a variable, or if one term is an injection
+applied to an overloaded symbol and the other is an injection applied to a
+variable, we handle these cases via the overload simplifier, which returns one
+of four cases:
+
+* Clash: unification fails
+* Simple resolution: decompose the two terms returned by the simplifier.
+* Narrowing resolution: constrain the solution with the constraint returned by
+  the simplifier and decompose the terms returned.
+
+Otherwise if one is an injection and the other is an overloaded symbol, fail.
+
+Map
+---
+
+If both patterns are of a map sort, we first normalize each map by categorizing
+each subterm according to one of three categories:
+
+1. Elements
+2. Variables
+3. Functions (other than `.Map`, `_Map_`, and `|->`)
+
+This process may yield the value \bottom for either map, in which case, fail.
+
+First, we remove all subterms present in both maps from each map.
+
+We then consider 2 cases:
+
+1. Either map still contains functions. In this case, constrain with an
+   equality between the two maps.
+2. Both maps contain no functions. In this case, apply AC unification to the
+   two maps.
```

Comment:
I think this needs some examples because I think I don't understand what "contains functions" means.